### ISSUE TITLE
Create Transparency on XOR/OR Gateway Completeness

### ISSUE THREAD

**Issue by till-stadtler** [camunda/camunda-modeler#4044]
### Problem you would like to solve

XOR and OR gateways raise incidents:
* XOR and OR gateways without a default when no sequence flow condition evaluates to true
* XOR gateways when multiple sequence flow conditions evaluate to true

I call gateways that exactly cover their logical output space "complete". Meaning, for all possible values of all involved process variables, there is no possibility to raise an incident. The sequence flow conditions on XOR and OR gateways must be configured as such that at least one conditions evaluates to true, or a default sequence flow exists. In addition, for XOR gateways, the conditions need to be mutually exclusive, such that no two conditions can be true simultaneously.

There is no tool support to check for this completeness. Additionally, the conditions on sequence flows behind gateways can not be seen in a single view, see #4043.
Adding tool support for this will improve the quality of the BPMN process and improve operations by
* inducing the addition of a default flow, if applicable
* reducing the number of involved process variables (reducing the logical output space)
* creating transparency under which conditions an incident will be raised

### Proposed solution

It is possible to create expressions that describe if an incident will be raised. These expressions become quite complicated if we also check for the existence and the data type of the involved process variables (incident when comparing String to Number,...). If we assume that all involved process variables exist and have the correct data type, the conditions would look like this:
```
OR gateway (check if default flow exists or at least one condition evaluates to true)
not(doesDefaultFlowExist or condition1 or ... or conditionN)

XOR gateway (check if default flow exists or at least one condition evaluates to true and if at least one condition is true, check that only one condition is true)
not(doesDefaultFlowExist or condition1 or ... or conditionN) or (condition1 or ... or conditionN) and not(condition1 and not(condition2) and ... and not(conditionN) or not(condition1) and condition2 and ... and not(conditionN) or ... or not(condition1) and not(condition2) and ... and conditionN)
```

For complete gateways, these expressions will always simplify to `false`, regardless of involved process variables. For incomplete gateways, these conditions express when an incident will be raised. To make such an expression useful, it should be simplified as much as possible by applying logical rules such as `not(A) or A = true`, etc.

I would like to see the resulting, simplified expression as a non-editable field in the property panel of the gateway. This will create transparency about the quality of the set conditions and set the expectations for operations when to expect incidents (in addition to non-existence and wrong data type).

Examples:
```
Complete XOR gateway:
default flow
condition 1: someString = "hello"
not(doesDefaultFlowExist or condition1) or (condition1) and not(condition1 or not(condition1)) 
= not(true or condition1) or condition1 and not(true) 
= not(true) or condition1 and false 
= false or false = false -> complete

Incomplete XOR gateway with potentially two matches:
default flow:
condition1: someString = "hello"
condition2: someOtherString = "world"
not(doesDefaultFlowExist or condition1 or condition2) or (condition1 or condition2) and not(condition1 and not(condition2) or not(condition1) and condition2)
= not(true or condition1 or condition2) or condition1 and condition2
= not(true) or condition1 and condition2
= false or condition1 and condition2
= condition1 and condition2 -> incomplete
// this final expression is useful to foster the understanding that an incident will be raised if someString = "hello" and someOtherString = "world"!

Incomplete XOR gateway with potentially no match:
condition1: someString = "hello"
condition2: someOtherString = "world"
not(doesDefaultFlowExist or condition1 or condition2) or (condition1 or condition2) and not(condition1 and not(condition2) or not(condition1) and condition2)
= not(false or condition1 or condition2) or condition1 and condition2
= not(condition1 or condition2) or condition1 and condition2
= not(condition1) and not(condition2) or condition1 and condition2 -> incomplete
// this final expression is useful to foster the understanding that an incident will be raised if neither or both expressions evaluate to true

Complete OR gateway:
condition1: someString = "hello"
condition2: someString = "world"
condition3: someString != "hello" and someString != "world" (effectively the default flow)
not(doesDefaultFlowExist or condition1 or ... or conditionN)
= not(false or condition1 or condition2 or condition 3)
= not(condition1 or condition2 or not(condition1) and not(condition2))
= false -> complete
// here, it is important to notice that condition 3 need to be broken down and related to condition1 and condition2 for the simplification to proceed

Incomplete OR gateway:
condition1: someString = "hello"
condition2: someString = "world"
not(doesDefaultFlowExist or condition1 or ... or conditionN)
= not(false or condition1 or condition2)
= not(condition1 or condition2) -> incomplete
// this might prompt the user to add a default flow (error event -> business reaction), or raise the awareness of when to expect an incident to be raised.
```

### Alternatives considered

The proposed solution needs an algorithm to simplify the calculated expressions. It might be easier to calculate a "truth table" with all involved process variables and their expressions (while understanding that expressions can be related) and calculate the final expression from there.

**Comment by nikku** [comment]
Thanks for opening this issue. Given we _had_ full understanding of the variable domains (data types, instances) used _then_ we [could use validation](https://github.com/bpmn-io/bpmnlint) to check if existing conditions are complete.

----

I don't regard this as something that we can support short term in a meaning full manner for two reasons:

* We'd need to be able to fully type variables:
   * data types: Is a variable a `string` or `number`?
   * instances: What values can a particular string have (i.e. `dish={ "salat" |  "pizza" }`)

**Comment by till-stadtler** [comment]
Hi @nikku,
I am not fully convinced that we need to know the full variable domain to calculate the condition when a gateway raises an incident. As shown above, this condition is calculated from all sequence flow conditions behind a gateway (plus the existence of a default flow).

I fully expect that most gateways without default flows that are used in real-life are incomplete. And that is totally fine. We want these gateways to raise an incident if a variable has a wrong data type or has a value that is not planned for in the gateway.

Here is a prime example of an incomplete gateway. The "Incident condition" creates transparency about when an incident will be raised:
![gateway_completeness_example](https://github.com/camunda/camunda-modeler/assets/117343024/358c9b8b-487f-4298-98ef-00af30d4d4de)
Explanation:
Because a comparison is used with a string, we know that `dish` needs to be of type `string`. Additionally, if dish is not either "salat" or "pizza", and incident will be raised. (There are some logical simplifications I missed above, but it does not invalidate the approach).

The visual representation of this "Incident condition" can be improved, as it is not-editable. For example, if there are conditions on data types, they can be shown separately. Using a list, we have an implicit "or" between the different items:
![gateway_completeness_example_2](https://github.com/camunda/camunda-modeler/assets/117343024/cad1fd27-4354-4856-ac59-90c73738da13)

A very complicated gateway might involve multiple variables and might result in a long list of "Incident conditions". This will prompt the user to simplify the gateway, e.g., by reducing the number of involved variables, introducing a default flow, etc.

**Comment by nikku** [comment]
@till-stadtler Thanks for the additional explanation. 

I agree that we can offer the user _additional help_, i.e. make the variable domain that _we assume_ more obvious as well as conditions.

What we cannot do is _assert_, and hence raise an error due to _incompleteness_.

**Comment by nikku** [comment]
I think this could be a great example of an "advanced analytics" plug-in that we add to the modeler to offer additional debugging information.

**Comment by nikku** [comment]
Moving this to `backlog` for the time being.

**Comment by till-stadtler** [comment]
@nikku,
I agree with your assessment that it would be a nice tool for advanced debugging!
Looking forward to pick this topic up again some time in the future!
